[CI] MetadataIndexStateServiceBatchingTests testBatchCloseIndices failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/vn2vj6dmamixc/tests/:server:test/org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests/testBatchCloseIndices


**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests.testBatchCloseIndices" -Dtests.seed=A054AE00F85B4FB8 -Dtests.locale=hi -Dtests.timezone=Pacific/Majuro -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
[Failure dashboard for `org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests#testBatchCloseIndices`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testBatchCloseIndices'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.Exception: Test abandoned because suite timeout was reached.

  at __randomizedtesting.SeedInfo.seed([A054AE00F85B4FB8]:0)

```

Actual exception:
```
May 30, 2024 8:26:54 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException	
WARNING: Uncaught exception in thread: Thread[#362,elasticsearch[node_s_0][clusterApplierService#updateTask][T#1],5,TGRP-MetadataIndexStateServiceBatchingTests]	
java.lang.AssertionError: 	
Expected: one of {<0>, <3>}	
     but: was <1>	
	at __randomizedtesting.SeedInfo.seed([A054AE00F85B4FB8]:0)	
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)	
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)	
	at org.elasticsearch.test.ESTestCase.assertThat(ESTestCase.java:2119)	
	at org.elasticsearch.cluster.metadata.MetadataIndexStateServiceBatchingTests.lambda$closedIndexCountListener$7(MetadataIndexStateServiceBatchingTests.java:213)	
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListener(ClusterApplierService.java:560)	
	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:546)	
	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:505)	
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:429)	
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:154)	
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:917)	
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:217)	
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:183)	
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)	
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)	
	at java.base/java.lang.Thread.run(Thread.java:1583)	
```

--- a/server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/tasks/ListTasksIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/tasks/ListTasksIT.java
@@ -29,10 +29,8 @@ import org.elasticsearch.transport.TransportService;
 
 import java.util.Collection;
 import java.util.List;
-import java.util.concurrent.BrokenBarrierException;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
 
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.empty;
@@ -102,7 +100,7 @@ public class ListTasksIT extends ESSingleNodeTestCase {
             }));
 
         // briefly fill up the management pool so that (a) we know the wait has started and (b) we know it's not blocking
-        flushThreadPool(threadPool, ThreadPool.Names.MANAGEMENT);
+        flushThreadPoolExecutor(threadPool, ThreadPool.Names.MANAGEMENT);
 
         final var getWaitFuture = new PlainActionFuture<Void>();
         clusterAdmin().prepareGetTask(task.taskId()).setWaitForCompletion(true).execute(getWaitFuture.delegateFailure((l, getResult) -> {
@@ -125,16 +123,6 @@ public class ListTasksIT extends ESSingleNodeTestCase {
         testActionFuture.get(10, TimeUnit.SECONDS);
         listWaitFuture.get(10, TimeUnit.SECONDS);
         getWaitFuture.get(10, TimeUnit.SECONDS);
-    }
-
-    private void flushThreadPool(ThreadPool threadPool, String executor) throws InterruptedException, BrokenBarrierException,
-        TimeoutException {
-        var maxThreads = threadPool.info(executor).getMax();
-        var barrier = new CyclicBarrier(maxThreads + 1);
-        for (int i = 0; i < maxThreads; i++) {
-            threadPool.executor(executor).execute(() -> safeAwait(barrier));
-        }
-        barrier.await(10, TimeUnit.SECONDS);
     }
 
     @Override

--- a/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java
+++ b/server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java
@@ -19,13 +19,13 @@ import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.core.CheckedRunnable;
 import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.CyclicBarrier;
-import java.util.concurrent.TimeUnit;
 
 import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_BLOCKS_WRITE_SETTING;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
@@ -126,6 +126,9 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
         // wait for the queue to have the second close tasks (the close-indices tasks)
         assertBusy(() -> assertThat(findPendingTasks(masterService, "close-indices"), hasSize(2)));
 
+        // wait for all ongoing tasks to complete on GENERIC to ensure that the batch is fully-formed (see #109187)
+        flushThreadPoolExecutor(getInstanceFromNode(ThreadPool.class), ThreadPool.Names.GENERIC);
+
         block2.run(); // release block
 
         // assert that the requests were acknowledged
@@ -208,14 +211,14 @@ public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase
     private static CheckedRunnable<Exception> blockMasterService(MasterService masterService) {
         final var executionBarrier = new CyclicBarrier(2);
         masterService.createTaskQueue("block", Priority.URGENT, batchExecutionContext -> {
-            executionBarrier.await(10, TimeUnit.SECONDS); // notify test thread that the master service is blocked
-            executionBarrier.await(10, TimeUnit.SECONDS); // wait for test thread to release us
+            safeAwait(executionBarrier); // notify test thread that the master service is blocked
+            safeAwait(executionBarrier); // wait for test thread to release us
             for (final var taskContext : batchExecutionContext.taskContexts()) {
                 taskContext.success(() -> {});
             }
             return batchExecutionContext.initialState();
         }).submitTask("block", new ExpectSuccessTask(), null);
-        return () -> executionBarrier.await(10, TimeUnit.SECONDS);
+        return () -> safeAwait(executionBarrier);
     }
 
     private static ClusterStateListener closedIndexCountListener(int closedIndices) {

--- a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
@@ -75,6 +75,7 @@ import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.common.util.CollectionUtils;
 import org.elasticsearch.common.util.Maps;
 import org.elasticsearch.common.util.MockBigArrays;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
 import org.elasticsearch.common.util.concurrent.EsExecutors;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
@@ -2290,12 +2291,30 @@ public abstract class ESTestCase extends LuceneTestCase {
         }
     }
 
+    /**
+     * Wait for all tasks currently running or enqueued on the given executor to complete.
+     */
     public static void flushThreadPoolExecutor(ThreadPool threadPool, String executorName) {
         final var maxThreads = threadPool.info(executorName).getMax();
         final var barrier = new CyclicBarrier(maxThreads + 1);
         final var executor = threadPool.executor(executorName);
         for (int i = 0; i < maxThreads; i++) {
-            executor.execute(() -> safeAwait(barrier));
+            executor.execute(new AbstractRunnable() {
+                @Override
+                protected void doRun() {
+                    safeAwait(barrier);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    fail(e, "unexpected");
+                }
+
+                @Override
+                public boolean isForceExecution() {
+                    return true;
+                }
+            });
         }
         safeAwait(barrier);
     }

--- a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
@@ -2290,6 +2290,16 @@ public abstract class ESTestCase extends LuceneTestCase {
         }
     }
 
+    public static void flushThreadPoolExecutor(ThreadPool threadPool, String executorName) {
+        final var maxThreads = threadPool.info(executorName).getMax();
+        final var barrier = new CyclicBarrier(maxThreads + 1);
+        final var executor = threadPool.executor(executorName);
+        for (int i = 0; i < maxThreads; i++) {
+            executor.execute(() -> safeAwait(barrier));
+        }
+        safeAwait(barrier);
+    }
+
     protected static boolean isTurkishLocale() {
         return Locale.getDefault().getLanguage().equals(new Locale("tr").getLanguage())
             || Locale.getDefault().getLanguage().equals(new Locale("az").getLanguage());

